deepreg/model/loss/util.py

```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         :return: negated loss.
         """
         return -super().call(y_true=y_true, y_pred=y_pred)
+
+
+EPS = tf.keras.backend.epsilon()
+
+
+def rectangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D rectangular
+    kernel for LocalNormalizedCrossCorrelation.
+
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
+        - kernel_vol, scalar indicating the sum of the coefficients of the equivalent
+                      3D kernel used for normalization purposes
+    """
+
+    kernel = tf.ones(shape=(kernel_size, 1, 1), dtype="float32")
+    return kernel
+
+
+def triangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D triangular
+    kernel for LocalNormalizedCrossCorrelation.
+
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
+        - kernel_vol, scalar indicating the sum of the coefficients of the equivalent
+                      3D kernel used for normalization purposes
+    """
+    fsize = int((kernel_size + 1) / 2)
+    pad_filter = tf.constant(
+        [
+            [0, 0],
+            [int((fsize - 1) / 2), int((fsize + 1) / 2)],
+            [0, 0],
+        ]
+    )
+
+    f1 = tf.ones(shape=(1, fsize, 1), dtype="float32") / fsize
+    f1 = tf.pad(f1, pad_filter, "CONSTANT")
+    f2 = tf.ones(shape=(fsize, 1, 1), dtype="float32") / fsize
+
+    kernel = tf.nn.conv1d(f1, f2, stride=[1, 1, 1], padding="SAME")
+    kernel = tf.transpose(kernel, perm=[1, 2, 0])
+
+    return kernel
```

Comment:
Sorry I may being stupid, but this seems not support multi-channel input?

Reply:
> Sorry I may being stupid, but this seems not support multi-channel input?

You are not nor does it. We had multiple discussions to support multiple channel images, which will be constrained also by GPU memory, but decided to leave it until users require... @MattClarkson @acasamitjana correct me if any update/plan.

Reply:
for now our images and labels are 1 channel only? but maybe add some assertion to be clear?

Reply:
labels have multiple channels but sampling one in each iteration - this reminds me, we can do the a simple thing to support multi-channel input using sampling, although it is not the same using and MI, but the same for most other losses. I don't think there is a multi-channel CC, rather just compute CC for each channel then average?

Reply:
> > Sorry I may being stupid, but this seems not support multi-channel input?
> 
> You are not nor does it. We had multiple discussions to support multiple channel images, which will be constrained also by GPU memory, but decided to leave it until users require... @MattClarkson @acasamitjana correct me if any update/plan.

If we are supporting multi-channel in the future, should we only use gaussian distributed spatial-wise but not channel-wise. I don't think any channel should be weighted more than other channels because its located in the middle.

Reply:
No, you are right. I couldn't think of a case where multi-channel applied (with regards the LNCC calculation) but happy to include the multi-channel case if needed. In this case, it would be independently applied to each feature map or as a 4-D kernel?

Reply:
@acasamitjana @kate-sann5100 i'd say implement with 3-d kernel, documented, and finish this ticket. We should start a new ticket to properly investigate this issue - i need to do some reading for this.

Reply:
I believe 3-d kernel should be enough. All we need for multi-channel input is to `tf.reduce_sum(tensor, axis=-1, keepdims=True)`. One annoying thing for implementation is that Tensorflow does not support group convolution, so we have to do convolution on each channel and sum up.